arrow-fx-stm/src/main/kotlin/arrow/fx/stm/TMap.kt

```diff
@@ -8,6 +8,153 @@ fun <K, V> STM.newTMap(fn: (K) -> Int): TMap<K, V> = TMap(newHamt(), fn)
 fun <K, V> STM.newTMap(): TMap<K, V> = newTMap { it.hashCode() }
 fun <K, V> STM.newTMap(hash: Hash<K>): TMap<K, V> = newTMap { hash.run { it.hash() } }
 
+/**
+ * A [TMap] is a concurrent transactional implementation of a key value hashmap.
+ *
+ * Based on a Hash-Array-Mapped-Trie implementation. While this does mean that a read may take up to 5 steps to be resolved (depending on how
+ *  well distributed the hash function is), it also means that structural changes can be isolated and thus do not increase contention with
+ *  other transactions. This effectively means concurrent access to different values is unlikely to interfere with each other.
+ *
+ * > Hash conflicts are resolved by chaining.
+ *
+ * ## Creating a [TMap]
+ *
+ * Depending on whether or not you are in a transaction you can use either [STM.newTMap] or [TMap.new] to create a new [TMap].
+ *
+ * There are a few alternatives because [TMap] can be supplied a custom hash strategy. If no argument is given it defaults to [hashCode].
+ *
+ * ## Reading an element with a key
+ *
+ * Reading from a [TMap] can be done using either [STM.lookup] or its alias [STM.get].
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TMap
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tmap = TMap.new<Int, String>()
+ *   val result = atomically {
+ *     tmap[1] = "Hello"
+ *     tmap[2] = "World"
+ *
+ *     tmap.lookup(1) + tmap[2]
+ *   }
+ *   //sampleEnd
+ *   println("Result $result")
+ * }
+ * ```
+ *
+ * > If the key is not present [STM.lookup] will not retry, instead it returns `null`.
+ *
+ * ## Inserting a value
+ *
+ * Inserting can be done using either [STM.insert] or its alias [STM.set]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TMap
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tmap = TMap.new<Int, String>()
+ *   atomically {
+ *     tmap.insert(1, "Hello")
+ *     tmap[2] = "World"
+ *   }
+ *   //sampleEnd
+ * }
+ * ```
+ *
+ * Another option when adding elements is to use [STM.plus]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TMap
+ * import arrow.core.toT
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tmap = TMap.new<Int, String>()
+ *   atomically {
+ *     tmap + (1 to "Hello")
```

Comment:
Should these be += instead of just plus if it returns Unit?

Reply:
That would actually make more sense :+1:

Reply:
Changed it, it does look a lot better^^